core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -192,4 +228,78 @@ public synchronized T next() {
       return queue.poll();
     }
   }
+
+  private static class Task<T> implements Supplier<Optional<Task<T>>>, Closeable {
+    private final Iterable<T> input;
+    private final ConcurrentLinkedQueue<T> queue;
+    private final AtomicBoolean closed;
+    private final int approximateMaxQueueSize;
+
+    private Iterator<T> iterator = null;
+
+    Task(
+        Iterable<T> input,
+        ConcurrentLinkedQueue<T> queue,
+        AtomicBoolean closed,
+        int approximateMaxQueueSize) {
+      this.input = Preconditions.checkNotNull(input, "input cannot be null");
+      this.queue = Preconditions.checkNotNull(queue, "queue cannot be null");
+      this.closed = Preconditions.checkNotNull(closed, "closed cannot be null");
+      this.approximateMaxQueueSize = approximateMaxQueueSize;
+    }
+
+    @Override
+    public Optional<Task<T>> get() {
+      try {
+        if (iterator == null) {
```

Comment:
nit: why not just initialize the iterator in the constructor?

Reply:
Tasks are constructed lazily, when iterating `ParallelIterator#tasks`.
Doing `tasks.hasNext()` likely will create a new Task (without returning it).
If we init iterator in the Task constructor, we would have Tasks that need closing, but we wouldn't be closing them.
See also https://github.com/apache/iceberg/pull/10691#discussion_r1677178824